### ISSUE TITLE
Null reference exception in InversePropertyAttributeConvention.ProcessNavigationAdded

### ISSUE THREAD

**Issue by mscrivo** [aspnet/EntityFrameworkCore#18030]
After upgrading our codebase from EF Core 2.2 to 3.0, we're encountering the following exception on any operation that uses the DBContext:

```
System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.InversePropertyAttributeConvention.ProcessNavigationAdded(IConventionRelationshipBuilder relationshipBuilder, IConventionNavigation navigation, InversePropertyAttribute attribute, IConventionContext`1 context)
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.NavigationAttributeConventionBase`1.ProcessNavigationAdded(IConventionRelationshipBuilder relationshipBuilder, IConventionNavigation navigation, IConventionContext`1 context)
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.ImmediateConventionScope.OnNavigationAdded(IConventionRelationshipBuilder relationshipBuilder, IConventionNavigation navigation)
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.RunVisitor.VisitOnNavigationAdded(OnNavigationAddedNode node)
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.OnNavigationAddedNode.Accept(ConventionVisitor visitor)
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.ConventionVisitor.Visit(ConventionNode node)
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.ConventionVisitor.VisitConventionScope(ConventionScope node)
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.ConventionBatch.Run()
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.ConventionBatch.Dispose()
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.ImmediateConventionScope.OnModelInitialized(IConventionModelBuilder modelBuilder)
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.OnModelInitialized(IConventionModelBuilder modelBuilder)
   at Microsoft.EntityFrameworkCore.Metadata.Internal.Model..ctor(ConventionSet conventions)
   at Microsoft.EntityFrameworkCore.ModelBuilder..ctor(ConventionSet conventions)
   at Microsoft.EntityFrameworkCore.Infrastructure.ModelSource.CreateModel(DbContext context, IConventionSetBuilder conventionSetBuilder)
   at Microsoft.EntityFrameworkCore.Infrastructure.ModelSource.<>c__DisplayClass5_0.<GetModel>b__1()
   at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
```

### Steps to reproduce

Unfortunately I don't have a minimal reproduction because I don't know what in particular in our large DB context is causing the issue.  2 out of 3 or contexts do work fine, it's just the most complex of them that doesn't.  There are hundreds of entities in this context, along with probably 50 query types (or keyless types I should say).   I tried stepping into the EF core code when the exception happens, hoping to get a clue of what in our context is bad, but with no luck.

### Further technical details

EF Core version: 3.0
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework:  .NET Core 3.0
Operating system: Windows 10 1903
IDE: Visual Studio 2019 16.3

**Comment by smitpatel** [comment]
Given stacktrace, it would likely be happening when adding EntityType for DbSet. Can you try adding entityTypes one-by-one to see which one causes the issue?

**Comment by mscrivo** [comment]
I can try, was hoping to avoid that as it's going to be extremely time consuming as we have hundreds of entities. Was hoping that the stacktrace would trigger someone's memory of what changed in this area that could help narrow it down.

On a side note, is sourceLink setup for the EF core libs? Debugging into the EF Core code would likely help me figure out what the issue is here, but when I try it, it complains that the source available on the source server does not match the binary and thus I can't step into any of the EF Core code.

**Comment by mscrivo** [comment]
Magically, I can step into the EF Core code today!  Source must have been updated since yesterday. Anyhow, I've identified the offending Entity, will report back once I have more info.

**Comment by mscrivo** [comment]
In `ProcessNavigationAdded` there is this block of code:

```C#
var newRelationship = ConfigureInverseNavigation(
    navigation.DeclaringEntityType.Builder,
    navigation.GetIdentifyingMemberInfo(),
    navigation.GetTargetType().Builder,
    attribute);
if (newRelationship != relationshipBuilder)
{
    var newNavigation = navigation.IsDependentToPrincipal()
        ? newRelationship.Metadata.DependentToPrincipal
        : newRelationship.Metadata.PrincipalToDependent;

    context.StopProcessingIfChanged(newNavigation);
}

The NRE is happening is happening when dereferencing newNavigation.Metadata because newNavigation is null in this case.  Seems like there is a missing null check here because `ConfigureInverseNavigation` has a couple cases where it explicitly returns null.

Still trying to figure out what about this particular entity is causing the problem. All I can see so far is that its a type that participates in a inheritance structure.  It's a type that inherits from an abstract base class, but it itself has no extra properties on it.
```

**Comment by mscrivo** [comment]
Ok figured it out:

We have a structure like this:

```C#
public abstract class FlaggedData {
    ForeignKey("UserID")]
    [InverseProperty("FlaggedDataUsers")]
    public User User { get; set; }

    [ForeignKey("CuratingUserID")]
    [InverseProperty("FlaggedDataCuratingUsers")]
    public User CuratingUser { get; set; }	

    ... other props
```
and then a couple of classes inheriting from that with no additional InverseProperties.  If I comment out those 2 inverse properties in the base, I no longer get the NRE.

**Comment by mscrivo** [comment]
@smitpatel Any further thoughts on this? seems to be a bug in 3.0 AFAICT?

**Comment by AndriySvyryd** [comment]
Fixed in 1c5e3089b09c786ada8bab7eeee1a81a64cc062e
